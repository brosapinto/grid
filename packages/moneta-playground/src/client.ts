import { ApolloClient, InMemoryCache, HttpLink, split } from "@apollo/client";
import { getMainDefinition } from "@apollo/client/utilities";
import { setContext } from "@apollo/client/link/context";
import { WebSocketLink } from "@apollo/client/link/ws";

// HTTP transport Link
const httpLink = new HttpLink({
  uri: "/graphql",
});

// WebSocket transport link
const wsLink = new WebSocketLink({
  uri: `ws://frontapi.qa-rows.com/ws`,
  options: {
    reconnect: true,
    lazy: true,
    connectionParams: {
      auth: sessionStorage.getItem("authToken"),
    },
    connectionCallback: (error) => {
      console.error("caput", error);
    },
  },
});

/**
 * Use WebSockets for subcriptions only, HTTP for everything else
 */
const transportLink = split(
  ({ query }) => {
    const definition = getMainDefinition(query);
    return (
      definition.kind === "OperationDefinition" &&
      definition.operation === "subscription"
    );
  },
  wsLink,
  httpLink
);

const authLink = setContext((_, { headers }) => {
  // get the authentication token from local storage if it exists
  const token = sessionStorage.getItem("authToken");
  // return the headers to the context so httpLink can read them
  return {
    headers: {
      ...headers,
      Authorization: token,
    },
  };
});

export default new ApolloClient({
  link: authLink.concat(transportLink),
  cache: new InMemoryCache(),
});

export { ApolloProvider } from "@apollo/client";
