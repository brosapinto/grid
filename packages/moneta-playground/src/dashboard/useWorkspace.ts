import { gql, useQuery } from "@apollo/client";

export const WORKSPACE = gql`
  query Workspace($id: ID!) {
    workspace: getWorkspace(workspaceId: $id) {
      id
      name
      slug
      folders {
        id
        apps {
          workspaceId
          id
          name
          modifiedAt
        }
      }
    }
  }
`;

export default function useWorkspace(id: string) {
  return useQuery<{ workspace: Workspace }, { id: string }>(WORKSPACE, {
    variables: { id },
  });
}
