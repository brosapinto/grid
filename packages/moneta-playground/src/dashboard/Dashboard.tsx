import React from "react";

import CreateApp from "./CreateApp";
import AppList from "./AppList";
import { RouteComponentProps } from "react-router-dom";
import useWorkspace from "./useWorkspace";

type DashboardProps = RouteComponentProps<{ wId: string }>;

const Dashboard: React.FC<DashboardProps> = (props) => {
  const { wId } = props.match.params;
  const { data, loading, error } = useWorkspace(wId);

  if (error) return <h2 data-testid="error">Oops, something went wrong!</h2>;
  if (loading || !data) return <p>Loading...</p>;

  const [folder] = data.workspace.folders ?? [{ apps: [] }];

  return (
    <div>
      <h1>Dashboard</h1>
      <CreateApp wId={wId} />
      <AppList apps={folder.apps} />
    </div>
  );
};

export default Dashboard;
