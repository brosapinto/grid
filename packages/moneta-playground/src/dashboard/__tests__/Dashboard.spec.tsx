import { MockedProvider } from "@apollo/client/testing";
import { render, cleanup } from "@testing-library/react";

import { WORKSPACE } from "../useWorkspace";
import Dashboard from "../Dashboard";
import { MemoryRouter, Route } from "react-router-dom";

jest.mock("../ToggleSort", () => () => <p>Toggle</p>);
jest.mock("../AppItem", () => (props: { app: App }) => (
  <tr>
    <td>{props.app.name}</td>
  </tr>
));

/*
**ATTENTION**

Even though GraphQL operations are mocked, they're asynchronous.
Therefore, it's necessary to wait for the next tick for assertions

The default state, when we don't wait, it's the _loading state_
*/

afterEach(() => cleanup());

it("starts by rendering a loading state", () => {
  const mocks = [{ request: { query: WORKSPACE }, result: { data: {} } }];
  const { container } = render(
    <MemoryRouter initialEntries={["/wId"]}>
      <MockedProvider mocks={mocks}>
        <Route path="/:wId" component={Dashboard} />
      </MockedProvider>
    </MemoryRouter>
  );

  expect(container).toMatchSnapshot();
});

it("renders an error when query fails", async () => {
  const mocks = [{ request: { query: WORKSPACE }, error: new Error("Ooop") }];

  const { findByTestId } = render(
    <MemoryRouter initialEntries={["/wId"]}>
      <MockedProvider mocks={mocks}>
        <Route path="/:wId" component={Dashboard} />
      </MockedProvider>
    </MemoryRouter>
  );

  const errorElem = await findByTestId("error");
  expect(errorElem).toMatchSnapshot();
});

it("renders a list of apps", async () => {
  const mocks = [
    {
      request: { query: WORKSPACE, variables: { id: "wId" } },
      result: {
        data: {
          workspace: {
            id: "wId",
            name: "wId",
            slug: "w-id",
            folders: [
              {
                id: "fId",
                apps: [
                  { workspaceId: "wId", id: 1, name: "app 1", modifiedAt: 0 },
                  { workspaceId: "wId", id: 2, name: "app 2", modifiedAt: 0 },
                ],
              },
            ],
          },
        },
      },
    },
  ];
  const { findByTestId } = render(
    <MemoryRouter initialEntries={["/wId"]}>
      <MockedProvider mocks={mocks}>
        <Route path="/:wId" component={Dashboard} />
      </MockedProvider>
    </MemoryRouter>
  );

  const appList = await findByTestId("apps");
  expect(appList).toMatchSnapshot();
});
