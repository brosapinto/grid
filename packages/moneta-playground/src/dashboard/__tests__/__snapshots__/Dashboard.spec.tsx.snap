// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a list of apps 1`] = `
<table
  data-testid="apps"
>
  <thead>
    <tr>
      <th>
        <p>
          Toggle
        </p>
      </th>
      <th />
      <th />
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        app 1
      </td>
    </tr>
    <tr>
      <td>
        app 2
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`renders an error when query fails 1`] = `
<h2
  data-testid="error"
>
  Oops, something went wrong!
</h2>
`;

exports[`starts by rendering a loading state 1`] = `
<div>
  <p>
    Loading...
  </p>
</div>
`;
